rllib/env/multi_agent_episode.py

```diff
@@ -0,0 +1,718 @@
+                terminated. The default is `False`, i.e. the episode is ongoing.
+            is_truncated: Optional. A boolean, defining, if an environment is
+                truncated. The default is `False`, i.e. the episode is ongoing.
+            render_images: Optional. A list of RGB uint8 images from rendering
+                the environment; the images include the corresponding rewards.
+            agent_ids: Obligatory. A list of strings containing the agent ids.
+                These have to be provided at initialization.
+
+        Returns: A `MultiAgentEpisode` instance.
+        """
+
+        self.id_: str = id_ or uuid.uuid4().hex
+
+        # Agent ids must be provided if data is provided. The Episode cannot
+        # know how many agents are in the environment. Also the number of agents
+        # can grwo or shrink.
+        self._agent_ids: Union[List[str], List[object]] = (
+            [] if agent_ids is None else agent_ids
+        )
+
+        # The global last timestep of the episode and the timesteps when this chunk
+        # started.
+        self.t = self.t_started = (
+            t_started if t_started is not None else max(len(observations) - 1, 0)
+        )
+        # Keeps track of the correspondence between agent steps and environment steps.
+        # This is a mapping from agents to `IndexMapping`. The latter keeps
+        # track of the global timesteps at which an agent stepped.
+        self.global_t_to_local_t: Dict[str, List[int]] = self._generate_ts_mapping(
+            observations
+        )
+
+        # Note that all attributes will be recorded along the global timestep
+        # in an multi-agent environment. `SingleAgentEpisodes`
+        self.agent_episodes: MultiAgentDict = {
+            agent_id: self._generate_single_agent_episode(
+                agent_id,
+                agent_episode_ids,
+                observations,
+                actions,
+                rewards,
+                states,
+            )
+            for agent_id in self._agent_ids
+        }
+
+        # obs[-1] is the final observation in the episode.
+        self.is_terminated: bool = is_terminated
+        # obs[-1] is the last obs in a truncated-by-the-env episode (there will no more
+        # observations in following chunks for this episode).
+        self.is_truncated: bool = is_truncated
+        # RGB uint8 images from rendering the env; the images include the corresponding
+        # rewards.
+        assert render_images is None or observations is not None
+        self.render_images: Union[List[np.ndarray], List[object]] = (
+            [] if render_images is None else render_images
+        )
+
+    def concat_episode(self, episode_chunk: "MultiAgentEpisode") -> None:
+        """Adds the given `episode_chunk` to the right side of self.
+
+        This is used for concatenating ongoing episodes in the buffer with
+        ongoing episodes in the sampler.
```

Comment:
Let's remove this comment. We shouldn't assume thing as to how this API is used.

Let's add instead:
- docstring
- the rules for when it's ok to concatenate:
- - IDs are identical
- - timesteps match (`t` of `self` matches `t_started` of `episode_chunk`)